### PR TITLE
on-line hard mining

### PR DESCRIPTION
**Pull request by CoinCheung**
This is a demo of ohem. I have tested on our own dataset, and the map/mar get boosted roughly 1 point.

The original  paper use nms to eliminate overlapping rois, but the original paper use ss to generate proposals, while this repository relies on rpn to generate proposals. So I did not implement associated part.

If this modification has potential risks or errors, please point them out and I will fix it.

### CODE REVIEW

File: maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py [hunk 1]
```diff
@@ -105,16 +108,79 @@ def subsample(self, proposals, targets):
...
+                [0, 1, 2, 3], device=device)
+
+        box_loss = smooth_l1_loss(
+            box_regression[sampled_pos_inds_subset[:, None], map_inds],
+            regression_targets[sampled_pos_inds_subset],
+            reduction='none',
+            beta=1,
+        ).sum(dim=1, keepdim=True)
+        ohem_loss = classification_loss
+        ohem_loss[sampled_pos_inds_subset[:, None]] += box_loss
```
Comment (WeihongM) [inline maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py]: This operation maybe is wrong, which should be modifed into
`ohem_loss[sampled_pos_inds_subset[:, None]] = ohem_loss[sampled_pos_inds_subset[:, None]] + box_loss`

Comment (CoinCheung) [reply maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py]: Hi, 
Thanks for reviewing. Would you please explain why this is wrong? I just do not understand the difference.

Comment (WeihongM) [reply maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py]: I am not sure for this, I have tried your code above, but I find it fail to add the box_loss to the correspond classification loss at ohem.  Maybe you can debug whether it's the same case.
Actually I changed the code
`        ohem_loss = classification_loss.clone()
        ohem_loss[sampled_pos_inds_subset[:, None]] = ohem_loss[sampled_pos_inds_subset[:, None]] + box_loss
`

File: maskrcnn_benchmark/structures/segmentation_mask.py [hunk 2]
```diff
@@ -361,8 +361,9 @@ def __init__(self, polygons, size):
         self.polygons = []
         for p in polygons:
             p = PolygonInstance(p, size)
-            if len(p) > 0:
-                self.polygons.append(p)
+            #  if len(p) > 0:
```
Comment (fmassa) [inline maskrcnn_benchmark/structures/segmentation_mask.py]: Why do we have this change?

File: maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py (hunk 1 above)
Comment (fmassa) [reply maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py]: I would the very least not use an in-place operation, as it might break backpropagation depending on the operation.

You could replace it with something like what @WeihongM mentioned

File: maskrcnn_benchmark/structures/segmentation_mask.py (hunk 2 above)
Comment (CoinCheung) [reply maskrcnn_benchmark/structures/segmentation_mask.py]: Hi, 

I believe this modification anymore is no longer needed any more. I modified this because our json files have empty `segmentation` field in the `annotations` part with will break the logic here.

Comment (CoinCheung) [reply maskrcnn_benchmark/structures/segmentation_mask.py]: I have re-committed this pr, and new link is [here](https://github.com/facebookresearch/maskrcnn-benchmark/pull/845). If you please, we could continue communicating in that page :)

File: maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py (hunk 1 above)
Comment (CoinCheung) [reply maskrcnn_benchmark/modeling/roi_heads/box_head/loss.py]: Thanks for reviewing this. I have modified as you suggested.

**Comment by chucklu** [comment]
@CoinCheung   You can get the delete branch in fork back  through https://github.com/betaflight/betaflight/pull/8112#issuecomment-492081397

**Comment by LouieYang** [comment]
@CoinCheung Why the ohem_loss.size(0) compared with self.batch_size_per_image? It seems like ohem_loss includes all proposals from all batch images and maybe it should be torch.split to image-wise and then use top_k to select?

**Comment by WeihongM** [comment]
@CoinCheung    same question with Louie Yang, think this implementation have some problem.
